Builder error when init a strut in unsafe block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 22.04.1 LTS
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
V full version: V 0.3.0 0133ff2.5a834a2
Git version: git version 2.34.1
Git vroot status: weekly.2022.32-1-g5a834a2e-dirty (139 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/test/mod_safety/emptynil.v`
```v
module main

pub struct SomeError {
mut:
	msg string
}

fn main() {
	a := unsafe { &SomeError('yo') }
	println(a)
}

```

**What did you expect to see?**

I don't know what should be the vlang behavior in this case.


**What did you see instead?**
```
==================
/tmp/v_1000/emptynil.1062198105169962526.tmp.c:12174: error: cannot convert 'struct string' to 'struct main__SomeError *'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2432,7 +2432,7 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		}
 	} else if to_sym.kind == .struct_ && to_type.is_ptr() {
 		if from_sym.kind == .alias {
-			from_type = (from_sym.info as ast.Alias).parent_type.derive(from_type)
+			from_type = (from_sym.info as ast.Alias).parent_type.derive_add_muls(from_type)
 		}
 		if !from_type.is_int() && final_from_sym.kind != .enum_ && !from_type.is_pointer()
 			&& !from_type.is_ptr() {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2432,7 +2432,7 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		}
 	} else if to_sym.kind == .struct_ && to_type.is_ptr() {
 		if from_sym.kind == .alias {
-			from_type = (from_sym.info as ast.Alias).parent_type
+			from_type = (from_sym.info as ast.Alias).parent_type.derive(from_type)
 		}
 		if !from_type.is_int() && final_from_sym.kind != .enum_ && !from_type.is_pointer()
 			&& !from_type.is_ptr() {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2430,6 +2430,19 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 			ft := c.table.type_to_str(from_type)
 			c.error('cannot cast `$ft` to struct', node.pos)
 		}
+	} else if to_sym.kind == .struct_ && to_type.is_ptr() {
+		if from_sym.kind == .alias {
+			from_type = (from_sym.info as ast.Alias).parent_type
+		}
+		mut is_int_lit := false
+		if node.expr is ast.IntegerLiteral {
+			is_int_lit = true
+		}
+		if !is_int_lit && !from_type.is_pointer() && !from_type.is_ptr() {
+			ft := c.table.type_to_str(from_type)
+			tt := c.table.type_to_str(to_type)
+			c.error('cannot cast `$ft` to `$tt`', node.pos)
+		}
 	} else if to_sym.kind == .interface_ {
 		if c.type_implements(from_type, to_type, node.pos) {
 			if !from_type.is_ptr() && !from_type.is_pointer() && from_sym.kind != .interface_

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2434,11 +2434,7 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		if from_sym.kind == .alias {
 			from_type = (from_sym.info as ast.Alias).parent_type
 		}
-		mut is_int_lit := false
-		if node.expr is ast.IntegerLiteral {
-			is_int_lit = true
-		}
-		if !is_int_lit && !from_type.is_pointer() && !from_type.is_ptr() {
+		if !from_type.is_int() && !from_type.is_pointer() && !from_type.is_ptr() {
 			ft := c.table.type_to_str(from_type)
 			tt := c.table.type_to_str(to_type)
 			c.error('cannot cast `$ft` to `$tt`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2434,7 +2434,8 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		if from_sym.kind == .alias {
 			from_type = (from_sym.info as ast.Alias).parent_type
 		}
-		if !from_type.is_int() && !from_type.is_pointer() && !from_type.is_ptr() {
+		if !from_type.is_int() && final_from_sym.kind != .enum_ && !from_type.is_pointer()
+			&& !from_type.is_ptr() {
 			ft := c.table.type_to_str(from_type)
 			tt := c.table.type_to_str(to_type)
 			c.error('cannot cast `$ft` to `$tt`', node.pos)

--- a/vlib/v/checker/tests/cast_to_ref_struct_err.out
+++ b/vlib/v/checker/tests/cast_to_ref_struct_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/cast_to_ref_struct_err.vv:9:16: error: cannot cast `string` to `&SomeError`
+    7 |
+    8 | fn main() {
+    9 |     a := unsafe { &SomeError('yo') }
+      |                   ~~~~~~~~~~~~~~~~
+   10 |     println(a)
+   11 | }

--- a/vlib/v/checker/tests/cast_to_ref_struct_err.vv
+++ b/vlib/v/checker/tests/cast_to_ref_struct_err.vv
@@ -1,0 +1,11 @@
+module main
+
+pub struct SomeError {
+mut:
+	msg string
+}
+
+fn main() {
+	a := unsafe { &SomeError('yo') }
+	println(a)
+}

